### ISSUE TITLE
[Question] Is there a way to fetch assets from cache immediately when offline? and skip fetching from network until connection to the server is re-established?

### ISSUE THREAD

**Issue by ericj17** [TalAter/UpUp#100]
Hi there,

Is there a way to fetch from cache directly and not go to network, until connection to the server is re-established?

## Context
I have a long list of assets to be cached (includes js, html files (used to compose parts of a modal popup), css files, etc.). My web app will load an offline read only version of my app when disconnected. The user can open a detailed view (a modal popup) of selected data, the data model is stored in local storage. This all is working, but the offline page is very slow to load all the assets because for each asset, it is trying to fetch from the network first. Once network fetch fails and times out, then it will fetch from cache. Is there a way to fetch from cache directly and not go to network, until connection to the server is re-established?

## Your Environment
My web app is built on ASP.NET MVC and hosted on IIS.
I am able to successfully trigger the loading of an offline read only version of my app with upup.

Firefox: 49.0

Let me know if anything needs further clarification.

Thank you very much for this library and help,
Eric

**Comment by TalAter** [comment]
> This all is working, but the offline page is very slow to load all the assets because for each asset, it is trying to fetch from the network first. Once network fetch fails and times out, then it will fetch from cache. Is there a way to fetch from cache directly and not go to network, until connection to the server is re-established?

Does this happen when you are truly offline, or is it just a slow connection setting? If you are offline, the request from the server should return an error immediately, and not wait for a timeout... unless there is something specific to your app that somehow sets that long timeout.

It could be helpful if you were to include some code that reproduces this issue.

In any way, the offline state is detected by the fact that requests fail. Relying on the connection status as reported by the browser is not reliable.

**Comment by ericj17** [comment]
Hi TalAter, thanks for your quick response and sorry for my late response.

I realized I mixed up the concept of an Offline **Server** vs an Offline **Device**.
We've tested that if we have our device is offline, e.g. network adapter is disabled, then assets requests fail immediately and retrieved right away from cache.

In the case that my device has internet connection but the Server is offline, then I see the 2 second timeout delay until Connection Refused error is returned, which I suppose is expected.

I made a modification to UpUp.sw.js to set it into an "offline mode" if any asset fails to be fetched by the network. When in "offline mode", assets will be fetched from the cache immediately without going to network first. 

A poll is also setup to run when in "offline mode" to ping an endpoint on the server to check if it comes back on line. If the poll ever gets a response from the server, it will set it back to an "online mode" and future asset fetches will go through the network first, then fallback on the cache. I've added this endpoint as another settings parameter to UpUp's start function. Also, added a 'callback on reconnect' setting parameter that will be called when connection is restored..

This is not ideal, but in our case, all of our assets are hosted at a single server, so I think this will work for us in the time being.

Welcome your thoughts on this approach. Thanks again!
